pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,446 @@
+			log.Errorf("Error sending current state of DNS rules: %v", err)
+			cancel() // Cancel the context to close the stream
+		}
+	}()
+
+	log.Debugf("StreamPolicyState waiting for context to be done")
+	select {
+	case <-streamCtx.Done():
+		log.Info("Closing the stream")
+		s.DeleteStream(stream)
+		return streamCtx.Err()
+	case <-s.ctx.Done():
+		s.closeServer()
+		log.Info("StreamPolicyState done")
+		return s.ctx.Err()
+	}
+}
+
+// ReceiveDNSpolicesACK receives the DNS policies ACKs from the client
+// If the success is false, we can send cancel signal to the channel
+// in that case SDP will recreate the stream.
+func (s *FQDNDataServer) ReceiveDNSpolicesACK(stream standalonednsproxy.FQDNData_StreamPolicyStateServer) error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			s.closeServer()
+			log.Info("Stopping the stream")
+			return s.ctx.Err()
+		case <-stream.Context().Done():
+			log.Info("Stream context is finished, closing the stream")
+			return stream.Context().Err()
+		default:
+			update, err := stream.Recv()
+			if err != nil {
+				log.Errorf("Failed to receive update: %v", err)
+				return err
+			}
+			log.Debugf("Received update: %v", update)
+			requestId := update.GetRequestId()
+			_, ok := s.dnsMappingResult.Load(requestId)
+			if !ok {
+				log.Errorf("Received Message id not found %s", requestId)
+			} else {
+				log.Debugf("Received response for dns message id: %s", requestId)
+
+				// We can send cancel signal to the channel if the response code is not NO_ERROR,
+				// in that case SDP will recreate the stream.
+				responseCode := update.GetResponse()
+				if responseCode != standalonednsproxy.ResponseCode_RESPONSE_CODE_NO_ERROR {
+					log.Errorf("Failed to update DNS policies")
+					cancel, ok := s.streams.Load(stream)
+					if ok {
+						cancel()
+					}
+				}
+			}
+			// Delete the request from the map
+			s.dnsMappingResult.Delete(requestId)
+			log.Debugf("Deleted from local cache for dns message id: %s", requestId)
+		}
+	}
+}
+
+// NewServer creates a new FQDNDataServer which is used to handle the Standalone DNS Proxy grpc service
+func NewServer(endpointManager endpointmanager.EndpointManager, updateOnDNSMsg updateOnDNSMsgFunc) *FQDNDataServer {
+	ctx := context.Background()
+
+	s := &FQDNDataServer{
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		ctx:                 ctx,
+		streams:             lock.Map[standalonednsproxy.FQDNData_StreamPolicyStateServer, context.CancelFunc]{},
+		currentSnapshot:     make(map[identity.NumericIdentity]policy.SelectorPolicy),
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+	}
+
+	go func() {
+		<-s.ctx.Done()
+		log.Info("FQDN service context done, cleaning up resources")
+		s.cleanupStreams()
+		s.closeServer()
+	}()
+
+	return s
+}
+
+func convertToBytes(ips []net.IP) [][]byte {
+	var byteIps [][]byte
+	for _, ip := range ips {
+		byteIps = append(byteIps, []byte(ip.String()))
```

Comment:
I'm confused by this. Why are we converting IPs (which are inherently a byte slice) to string to then convert that string to bytes?

Reply:
Okay wasn't aware of that. I changed it so it does not convert it to string as IPs are stored as byte slice.